### PR TITLE
feat: add friendly captcha

### LINKED ISSUE #1598: Add Friendly-Captcha to Web v2
### Overview

We've decided to use friendly captcha for the legacy wallet migration.

We'll need to add Friendly Captcha to the legacy wallet login screen that will have to be completed and a solution sent to be verified on the server.

### References and additional details

https://docs.friendlycaptcha.com/#/

### Acceptance Criteria

A Friendly Captcha is used by the login screen before a user can attempt to login

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by theoboldfrazier**
## Description

Adds a friendly captcha to legacy login form.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1598 

## Risk

Low risk to legacy login, behind a feature flag.

## Testing

You should see a captcha automatically show and verify that you are a person. Once verified, the captcha disappears.

## Screenshots (if applicable)

Screen Shot 2022-04-29 at 14 57 46

### CODE REVIEW

**Review by cjthompson — APPROVED** [review]

File: src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx [hunk 1]
```diff
@@ -110,9 +115,12 @@ export const LegacyLogin = () => {
...
-  const onCaptchaRequested = (e: React.ChangeEvent<HTMLInputElement>) => {
-    // Check the captcha in case the captcha has been validated
-    setIsCaptchaSolved(e.target.checked)
+  const handleCaptcha = (solution: string) => {
+    if (typeof solution === 'string') {
+      setCaptchaSolution(solution)
+    } else {
+      setCaptchaSolution(null)
+    }
   }
```
Comment (0xdef1cafe) [inline src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx:124]:
```suggestion
```

File: src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx [hunk 2]
```diff
@@ -163,31 +171,13 @@ export const LegacyLogin = () => {
...
-                    isChecked={isCaptchaSolved}
-                  >
-                    {translate('common.notRobot')}
-                  </Checkbox>
-                </Card.Body>
-              </Card>
-              <FormErrorMessage>{errors.captchaChallenge?.message}</FormErrorMessage>
+            <FormControl isInvalid={!captchaSolution} mb={4} mt={6}>
+              <Box flexDir={'row'} justifyContent='center' alignItems={'center'}>
+                <FriendlyCaptcha handleCaptcha={handleCaptcha} />
```
Comment (0xdef1cafe) [inline src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx:176]:
```suggestion
                <FriendlyCaptcha handleCaptcha={setCaptchaSolution} />
```

File: src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx (hunk 1 above)
Comment (0xdef1cafe) [reply src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx:124]: also - this type sig is different to another component where it's `string | any` - which is it?

File: src/context/WalletProvider/NativeWallet/components/Captcha.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    },
+    [handleCaptcha],
+  )
+
+  const errorCallback = useCallback(
+    (err: any) => {
+      handleCaptcha(err)
+    },
+    [handleCaptcha],
+  )
```
Comment (0xdef1cafe) [inline src/context/WalletProvider/NativeWallet/components/Captcha.tsx:27]:
```suggestion
```

File: src/context/WalletProvider/NativeWallet/components/Captcha.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  useEffect(() => {
+    if (!widget.current && container.current) {
+      widget.current = new WidgetInstance(container.current, {
+        startMode: 'auto',
+        doneCallback: doneCallback,
+        errorCallback: errorCallback,
```
Comment (0xdef1cafe) [inline src/context/WalletProvider/NativeWallet/components/Captcha.tsx:34]:
```suggestion
        doneCallback: handleCaptcha,
        errorCallback: handleCaptcha,
```

File: src/context/WalletProvider/NativeWallet/components/Captcha.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      })
+    }
+
+    return () => {
+      if (widget.current !== undefined) widget.current.destroy()
+    }
```
Comment (0xdef1cafe) [inline src/context/WalletProvider/NativeWallet/components/Captcha.tsx:40]:
```suggestion
    return () => widget.current?.destroy?.()
```

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
few suggested simplifications - some stylistic

unless i'm missing something these callbacks seem entirely unnecessary?

File: src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx (hunk 1 above)
Comment (0xdef1cafe) [reply src/context/WalletProvider/NativeWallet/components/LegacyLogin.tsx:124]: if we just pass down `setCaptchaSolution` to the child it'll be referentially stable and we don't need the callbacks in the child either?

File: src/context/WalletProvider/NativeWallet/components/Captcha.tsx (hunk 3 above)
Comment (theoboldfrazier) [reply src/context/WalletProvider/NativeWallet/components/Captcha.tsx]: @0xdef1cafe Good find. Yeah, those callbacks were extraneous.

I pulled them out.

File: src/context/WalletProvider/NativeWallet/components/Captcha.tsx (hunk 5 above)
Comment (theoboldfrazier) [reply src/context/WalletProvider/NativeWallet/components/Captcha.tsx]: Done.

**Comment by 0xdef1cafe** [comment]
hey @theoboldfrazier - as discussed w/ @DiggyDiggy2 can we please center the captcha inside the modal? 🙏

**Comment by theoboldfrazier** [comment]
@0xdef1cafe done.

**Review by elmutt — DISMISSED** [review]

**Review by mrnerdhair — CHANGES_REQUESTED** [review]
Gimme a minute to double-check the `friendly-captcha` library itself and see what it does. I looked at it briefly a while ago, and I have a vague memory that we might need to put it in an iframe for some reason.

**Review by 0xdef1cafe — DISMISSED** [review]

**Comment by 0xdef1cafe** [comment]
@theoboldfrazier env vars added to cloudflare and fleek

**Comment by mrnerdhair** [comment]
Right, so looks like the `friendly-challenge` library sets `innerHTML` using some string-interpolated templates in order to render its UI. It doesn't look like it particularly _needs_ to do it that way -- it could just as easily do things in a provably-safe manner -- so I suggest we just make it work like that, probably using just `patch-package` for convenience.

Scope would be basically https://github.com/FriendlyCaptcha/friendly-challenge/blob/217438a5002e99134bbf0d50e097efb7b29082d0/src/captcha.ts, where the `get*HTML()` would need to return DOM elements; probably by importing `lit-html` (which they used `innerHTML` to [avoid](https://github.com/FriendlyCaptcha/friendly-challenge/blob/217438a5002e99134bbf0d50e097efb7b29082d0/src/dom.ts#L12) upstream since they were optimizing for lower dependency count).

Suggested AC would be that the modified `friendly-challenge` library works under a CSP of `trusted-types lit-html; require-trusted-types-for 'script'`. While `web` won't work under that policy out-of-the box (mainly because `react-dom` hasn't gotten their trusted types support out from behind a feature flag yet), it's a really easily-definable and easily-testable way to show that the library's performing safely.

I suggest that someone:
1) Spin up a blank app -- just `yarn init` in an empty folder, and install `microbundle`, `patch-package`, `postinstall-postinstall`, `lit-html`, and `friendly-challenge`. Set the build command to `microbundle` and the postinstall command to `patch-package`.
2) Cook up the world's stupidest static HTML page which imports the bundle `microbundle` spits out.
3) Write a tiny bit of code -- in plain JS b/c installing `tsc` wouldn't be worth the effort -- that imports the `friendly-challenge` library and renders the captcha to the page.
4) Make sure that all works, then add the CSP using a meta tag to break it.
5) Unbreak everything by screwing with `node_modules/friendly-challenge` until it uses `lit-html` instead of `innerHTML`
6) `patch-package friendly-challenge` to generate the patch
7) Cut-n-paste that patch into this PR

**Review by cjthompson — APPROVED** [review]

**Comment by cjthompson** [comment]
I don't know why the build is failing. It works just fine locally, even when doing a clean and installing from scratch.

**Review by mrnerdhair — APPROVED** [review]

**Pull request merged**